### ISSUE TITLE
Use Private Use Area for FL_Button and FL_key... constants

### ISSUE THREAD

**Issue by darealshinji** [fltk/fltk#58]
As indicated by the comment in `Enumerations.H`:
``` C
// FIXME: These codes collide with valid Unicode keys
```
Might be a good idea to make this optional so people don't accidentally break their programs?

Kinda like this maybe:
``` C
#ifdef SOMETHING_SOMETHING_PUA
#define FL_Button 0xE000
#else
#define FL_Button 0xFEE8
#endif
```

**Comment by erco77** [comment]
Yes -- and we apparently already use some of them in this section of Enumerations.H:

```
// These use the Private Use Area (PUA) of the Basic Multilingual Plane
// of Unicode. Guaranteed not to conflict with a proper Unicode character.

// These primarily map to the XFree86 keysym range
#define FL_Volume_Down  0xEF11   /* Volume control down        */
[..]
#define FL_Sleep        0xEF2F   /* Put system to sleep        */
#define FL_Favorites    0xEF30   /* Show favorite locations    */
```

..so we probably  just need to apply that for the 3 dozen or so 0xff?? codes above that (FL_Button, FL_BackSpace, etc)
Not sure what impact that may have on other parts of the code, but I guess if we're gonna change stuff like this, FLTK 4.0 would be the time to do it.

**Comment by ManoloFLTK** [comment]
I'm not sure that the collision between these values and Unicode is a problem.
These values are used to assign the global variable Fl::e_keysym with a number
expressing what key was last hit. Then, both library and user code can compare
Fl::e_keysym or its public accessor Fl::event_key() to specific symbolic values,
e.g., FL_BackSpace, FL_KP_xxx.

The numeric values in Enumerations.H match X11-defined keysyms
given in file X11/keysymdef.h, which won't change.
The X11 FLTK platform converts X11 keycodes into keysyms using XKeycodeToKeysym()
and that's known to work with most (all?) keyboards. 
The content of Fl::event_text() is produced by XUtf8LookpString() but the 
values under consideration (FL_BackSpace and friends) are to be used only in
relation to Fl::event_key(), not to Fl::event_text(). Text whose Unicode encoding
collides with the values under consideration is stored by FLTK in Fl::event_text()
not in Fl::event_key().

Other platforms (Windows and macOS) proceed by defining in Fl_System_Driver
a mapping between the OS-specific key encoding and FLTK's encoding.

In summary, the OS-specific naming of keys is transformed by class Fl_System_Driver
into an FLTK-controlled series of key codes listed in file Enumerations.H which
match key codes used by X11 to support keyboards. As long as X11 can support variant keyboards, FLTK should be able to support them with its present code.

**Comment by Albrecht-S** [comment]
Note: I took the freedom to edit Manolo's reply fixing a typo: XUtf8LookpString -> XUtf8LookupString for other readers.

**Comment by Albrecht-S** [comment]
I agree with Manolo. Further I'd like to mention that text in Fl::event_text() is UTF-8 encoded and does not collide with Unicode symbols in any way.

**Comment by Albrecht-S** [comment]
There's another related issue though [in the same file](https://github.com/fltk/fltk/blob/0d7a9e7f371b8a5f10bddae6fd8276bba283a9cf/FL/Enumerations.H#L538) concerning modifiers which reads:
```
// FIXME: it would be nice to have the modifiers in the upper 8 bit so that
//        a unicode key (24bit) can be sent as an unsigned with the modifiers.
```
The defined modifier values are used in conjunction with shortcuts that can be defined for menus and labels. Since these modifier values are above the lower 16 bits we can only combine 16-bit Unicode values with modifier keys for shortcuts.

As the comment says, it would be worthwile to change this and move the modifier bits to the upper 8 bits of a 32-bit `int` or (my comment:) at least above the lower 21 bits which would be enough to store all Unicode values. As a consequence of this move we'd have to move the FL_Button* values further down into the Unicode range but these wouldn't collide with the shortcut definition, hence this would IMHO be doable. I started working on a proof of concept (currently stalled but I hope to continue soon). I'm not sure yet how this works out. I'll let you know of my progress.

The problem I see with this is that we'd have to change named constants which is generally fine (that's what named constants are for, right?), but I'm not sure that **every** FLTK program out there uses the constant **_names_** correctly -- as opposed to self-defined or literal numeric values.

What do you think, can we rely on the constant **names** and redefine the modifier values?

**Comment by erco77** [comment]
I too wondered how there was a collision between keycodes and unicode, but given the FIXME and the OP's issue I figured it must somehow come up.

I'm left wondering then why the issue was reported: was there an actual collision encountered (and if so under what circumstance?) or is this just a request to follow up on a FIXME comment that is quite possibly no longer relevant?

(Matt added the OP "FIXME" comment in 2009 (git blame), so if there's lingering concern, we could ask him about it)

**Comment by darealshinji** [comment]
I didn't encounter any issues, no. It was because of the comment and I just wondered why the newer enum values were assigned to the PUA but not the old ones.

**Comment by ManoloFLTK** [comment]
I'm not sure to understand that part:
"As a consequence of this move we'd have to move the FL_Button* values further down into the Unicode range but these wouldn't collide with the shortcut definition, hence this would IMHO be doable. "

About "What do you think, can we rely on the constant names and redefine the modifier values?" 
I don't see why a user would have had the idea to redefine FL_SHIFT .
So I would believe it's very safe to modify the value given to FL_SHIFT and friends
in file FL/Enumerations.H

**Comment by Albrecht-S** [comment]
@ManoloFLTK: Part 1: we can use 32 bits (or 31 bits if we want to leave the sign bit alone) and we have 7 modifier bits. Hence we need to "move" the other (mouse) `FL_Button*` values into the bit range 16-23 which would be in the range of Unicode values but would never be combined with numerical values like `FL_ALT | FL_SHIFT | 'u'` as used in menu shortcuts.

Part 2: I didn't mean that a user redefined `FL_SHIFT` but that some user code might use the numerical value like `0x10000` instead of `FL_SHIFT`. Anyway, I also believe that user code **_should_** use the names of the constants rather than their numerical values.

**Comment by erco77** [comment]
A thing to watch out for is fluid - IIRC it sometimes (always?) uses keycodes instead of macro constants for e.g. shortcuts and such, but I haven't looked lately.
If need be we could make an expanded version of FL/names.h that includes all those macro names for keypresses,, so fluid can save macro names in the .fl files and code, instead of magic numbers.

**Comment by Albrecht-S** [comment]
Yep, good point. AFAICT fluid writes numeric values for "everything" into the .fl files and into the generated .cxx and .h files. I'll have to check that.

**Comment by ManoloFLTK** [comment]
A user might have done as has been done in FLTK
in function Fl::test_shortcut(unsigned) of file src/fl_shortcut.cxx
    if ((shortcut&shift) != (shortcut&0x7fff0000)) return 0;
after having noticed in Enumerations.H that all key modifiers
are covered by this mask: 0x7fff0000
This mask should have been written ~(FL_KEY_MASK)
but even FLTK did not do it.

**Comment by ManoloFLTK** [comment]
My understanding is that the value of FL_SHIFT and friends could be changed
to allow to use unicode values higher than 0xFFFF as keyboard shortcuts.
If that is correct, I wonder if keyboards with keys labelled with such characters exist?

**Comment by MatthiasWM** [comment]
I will put that on the list for 1.5.0 . There are going to be some bigger changes and chances with 1.5.0 going C++11 or C++17. This is too big of a change for 1.4.0.

**Comment by MatthiasWM** [comment]
Ok, I have tried to construct a case where this is important. Usually, the key codes have nothing to do with unicode values. Pressing Fl_Volume_Down should not generate unicode (not even 🔈↓ ). All keycodes use the lower 16 bits, and all modifiers use the upper 16 bits. 

The only time where we mix Unicode and modifiers is for shortcuts. This is not perfect. For example, a Gothic keyboard that has the letter Lagus "𐌻" could not differentiate between the unshifted Lagus `U-1033B` and the shifted Combining Square Below " ̻" `FL_SHIFT U+033B`.

That is not to say that there is *never* a conflict between maybe some emoji keys on a future keyboard, but for the time now, I fear that moving the shift key flags around will have a bad effect on some existing software (including everything created with Fluid).

I am closing this.
